text/0003-performance-data.md

```diff
@@ -0,0 +1,83 @@
+
+# Summary
+[summary]: #summary
+
+Currently there is Transport performance data and MLAPI performance data that is not being propagated up externally for any consumer to render or consume. The use cases for this would be useful for technologies such as a **Network Stats Overlay** or a **Profiler** (**Network Messages** and **Network Operation** need this data in the vanilla Unity Profiler). Users need a viable and low overhead way to obtain this data (such as RPCs sent this frame/tick) from MLAPI and utilize this data in the way that they want for their own metrics.
+
+# Motivation
+[motivation]: #motivation
+
+This is needed to provide both a default and user specific implementations for future Network Profilers and Network Stats overlays. The MLAPI source code can remain Open Source, and yet, the data being streamed can be hooked up to external and low level systems.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+In order to allow first and third party rendering of the data we will need to allow both MLAPI and its transports to propagate data up the chain. 
+If the profiler flags are enabled the following will need to happen: 
+Transports will need to output the number of bytes being sent in the current tick and the latency. The data will be written out to a ProfilerTickData chunk. This data will be generic and transport agnostic.
+MLAPI will need to output the number of networked objects and the amount of RPCs being sent. This data will be added to the ProfilerTickData chunk.
+
+External Tools and APIs will be able to then get this fully written ProfilerTickData at the end of the frame for them to do what they wish (Render to Profiler Stats Overlay or Display/Record in a Profiler or file)
+
+As this data is largely a struct like object, accessing its data will be as simple as accessing fields. Additionally the rendering or logging of these statistics will fall on the user and not affect the Tick or be done at all within the MLAPI framework.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+
+In order to limit the interface, we are pushing the collected set or performance data to be obtainable only at the end of a tick. The following is roughly what this object could look like. Currently this object will be updated in ```NetworkManager.LateUpdate()``` for the MLAPI and ```Transport.Send()``` and other relevant transport method for the transport layer. At the end of the tick the user will be notified that the relevant data is complete and ready (both the MLAPI data and Transport data).
+
+```cs
+class ProfilerTickData {
+    int tickID;
+
+    // transport-level
+    int bytesSent;
+    int bytesReceived;
+    int tickDuration;
+    ...
+
+    // mlapi-level
+    int numberOfRPCs;
+    ...
+
+    // potentially? For any remaining edge cases?
+    Dictionary<string,Object> kvStore;
```

Comment:
This could be useful but I'd leave it as a future possibility rather than part of the initial implementation

Reply:
I have it here as means to allow people to publicly comment on its possibility. I rather have it here with overwhelming sentiments of its removal then to have a large amount of people wonder why the implementation was so closed.

Reply:
I think this approach it mixing 2 other distinct approaches:

- be opinionated: type types into struct and pass it though
- be unopinionated: fill Dict with keys-values and pass it through

In my opinion, we should settle on one approach (either opinionated or unopinionated struct passthrough) and don't mix both of them.

In this scenario, I'm more inclined towards being unopinionated and just returning a Dict<string, int> back but that comes with its downsides:

- Dict<string, int> could be too limited — maybe everything can't be ints?
- Dict<string, object> introduces boxing overhead

I'd like to discuss more about this struct's layout, let me know what you guys think :)

Reply:
If thats the line we want to make then i favor typed types into the struct. There shouldn't be that many cases we aren't capable of thinking of. Most are known and any new ones can be added.

Reply:
in terms of API-design, I'm still not sure `Dict<string, object>` vs `struct OurCustomProfilingData` here. Dict approach would give us more flexibility with the cost of performance and less clue what will that Dict will contain — on the other hand, custom struct approach might be too opinionated and require lots of iterative changes over time.

I think we should discuss about this in a wider group, I will schedule a meeting for us :)

Reply:
We're going with prenamed fields in a struct. If anything we can add a dictionary to it at some later date. but more than likely we will just add the missing fields